pkg/preflight/validate_specs.go

```diff
@@ -4,45 +4,37 @@ import (
 	"reflect"
 
 	"github.com/replicatedhq/troubleshoot/pkg/apis/troubleshoot/v1beta2"
+	"github.com/replicatedhq/troubleshoot/pkg/loader"
 	"github.com/replicatedhq/troubleshoot/pkg/multitype"
 	"github.com/replicatedhq/troubleshoot/pkg/types"
 )
 
 // validatePreflight validates the preflight spec and returns a warning if there is any
-func validatePreflight(specs PreflightSpecs) *types.ExitCodeWarning {
+func validatePreflight(kinds *loader.TroubleshootKinds) *types.ExitCodeWarning {
 
-	if specs.PreflightSpec == nil && specs.HostPreflightSpec == nil && specs.UploadResultSpecs == nil {
+	if len(kinds.PreflightsV1Beta2) == 0 && len(kinds.HostPreflightsV1Beta2) == 0 {
 		return types.NewExitCodeWarning("no preflight or host preflight spec was found")
 	}
 
-	if specs.PreflightSpec != nil {
-		warning := validatePreflightSpecItems(specs.PreflightSpec.Spec.Collectors, specs.PreflightSpec.Spec.Analyzers)
+	for _, spec := range kinds.PreflightsV1Beta2 {
+		warning := validatePreflightSpecItems(spec.Spec.Collectors, spec.Spec.Analyzers)
 		if warning != nil {
 			return warning
 		}
 	}
 
-	if specs.HostPreflightSpec != nil {
-		warning := validateHostPreflightSpecItems(specs.HostPreflightSpec.Spec.Collectors, specs.HostPreflightSpec.Spec.Analyzers)
+	for _, spec := range kinds.HostPreflightsV1Beta2 {
+		warning := validateHostPreflightSpecItems(spec.Spec.Collectors, spec.Spec.Analyzers)
 		if warning != nil {
 			return warning
 		}
 	}
 
-	if specs.UploadResultSpecs != nil {
```

Comment:
can we avoid losing the validation of this type?

Reply:
> can we avoid losing the validation of this type?

TL;DR - We aren't losing the validation. It takes place [here](https://github.com/banjoh/troubleshoot/blob/7def396527932d1bf377c137c20769216c2849cd/pkg/preflight/validate_specs.go#L19-L24) where that list of `troubleshootv1beta2.Preflight` objects will contain both `PreflightSpec` and `UploadResultSpecs` preflight objects which were in the original `PreflightSpecs` struct

**Longer description**
Changes in this PR have gotten rid of the [PreflightSpecs struct](https://github.com/replicatedhq/troubleshoot/blob/d0d6f00c0f0a08d579fc4af498a007f049525d5a/pkg/preflight/read_specs.go#L14) in favour of using `TroubleshootKinds`.

Here is a comparison of the structs

```go
type PreflightSpecs struct {
	PreflightSpec     *troubleshootv1beta2.Preflight
	HostPreflightSpec *troubleshootv1beta2.HostPreflight
	UploadResultSpecs []*troubleshootv1beta2.Preflight
}
```

```go
type TroubleshootKinds struct {
	// redacted ...
	HostPreflightsV1Beta2   []troubleshootv1beta2.HostPreflight
	PreflightsV1Beta2       []troubleshootv1beta2.Preflight
}
```

As you can see both `PreflightSpec` and `UploadResultSpecs` store `*troubleshootv1beta2.Preflight`. I changed the logic avoid transforming `TroubleshootKinds` to `PreflightSpecs` in favour of just using `TroubleshootKinds` in an effort of having a common type that houses troubleshoot spec objects.

Changes can be found in
- Validation code
- Code to read spec inputs